### ISSUE TITLE
Refs to mysite dir maintained in composer.json

### ISSUE THREAD

**Issue by phptek** [silverstripe/recipe-plugin#6]
# Overview

I'm having trouble understanding how both the "public" and "mysite" dirs are created when using "type: "silverstripe-recipe" in my project's "composer.json". In lieu of the resolution to silverstripe/silverstripe-framework#7946 the recipe will create a "mysite" dir for me, even if I already have an "app" dir present. The only way to prevent this is to ensure I maintain the relevant entries in composer's "extra" block ala:

```
   "extra": {
        "project-files-installed": [
            "mysite/.htaccess",
            "mysite/_config.php",
            "mysite/_config/mysite.yml",
            "mysite/code/Page.php",
            "mysite/code/PageController.php"
        ],
```
...which if I'm not ever intending on using "mysite" - ever - seems a bit wasteful, just to prevent the recipe from creating the dir. Perhaps I'm missing something, otherwise at worst I'd call this a bug and at best, a wee bit of technical debt to have to maintain.

# Acceptance Criteria

 * As a dev with an app folder upgrading a 3.x install, I don't want to deal with recipe internals in order to stop it from copying files into the wrong place on any `composer update` call
 * If I have an `app` folder, any default files located in `mysite` of the recipe-core will be placed in the same directory structure in `app`

# Notes

 * We'll be using `app` going forward (most likely in 4.2) anyway
 * The upgrader notes if there's no `app` folder, separate issue
 * Assumption: Does not rely on shifting composer metadata into a `recipe.lock` file

**Comment by chillu** [comment]
> ... how both the "public" and "mysite" dirs are created

So just to be clear: You expect the `public` folder, but didn't expect the `mysite` folder, because you already have an `app` folder?

**Comment by robbieaverill** [comment]
Actually yeah, I've experienced this before too:

* Create a project locally
* Change mysite to app
* Build the application
* Deploy to web server and experience 500 errors
* Bang head on table, eventually copy and paste the installer file back in and run it
* It forces you to rename app back to mysite
* Installer completes, application works again
* Rename mysite back to app

I didn't spend any time looking into why this was happening, but that was my process for it with a 4.0 application.

**Comment by tractorcow** [comment]
> ...which if I'm not ever intending on using "mysite" - ever - seems a bit wasteful, just to prevent the recipe from creating the dir. 

@phptek yes it is a bit wasteful, however we absolutely need to register "installed" items in a place that respects file deletions. The fact that those records exist are essential to tell the recipe that those were installed AND if they no longer exist in that location, those were deleted intentionally and should NOT be reinstalled.

Removing the registration, unfortunately, will re-install those files the next time you update, because the recipe has no context for those files. It'll think "oh, that file was never installed into mysite, it must be a new file just added to the recipe, I'll go install it!".

You should be able to safely rename mysite to whatever you want, as long as you never delete those composer.json entries.

So tl;dr, yes we acknowledge it's wasteful, but it's intentional and necessary.

`public` folder is absolutely a different case. It is never ever created; You must either create it yourself, or install a project which has that folder created explicitly inside it.

5.x we will force a `public` folder for each recipe, which cannot be renamed or removed.

**Comment by tractorcow** [comment]
@phptek would you prefer that `project-files-installed` was written to a file alongside composer.json instead of inside it? It would be mandatory to commit this to source control, so the reason for inlining it was to prevent it being accidentally omitted.

**Comment by tractorcow** [comment]
E.g. `recipe.lock` :D

**Comment by tractorcow** [comment]
Not sure if can be done in a minor release... maybe major release only?

Marking as rfc so we can build this into an action plan for what we actually want to build.

**Comment by phptek** [comment]
@tractorcow `public` was introduced in a minor release (much to the consternation of our sysadmins).. :-P

**Comment by phptek** [comment]
@tractorcow re: `.recipe.lock` Yes, bang on what I was thinking as I was reading through. A separate cache of "installed" ("unwanted", "redundant", "not-clobbered" etc.) files. Thanks all.

**Comment by tractorcow** [comment]
Yeah, we can release new features in minor releases. Changing / removing features sometimes requires a new major version, unless you can do it an an extensible way or via opt-in.

I did have a think about this, and we could implement `.recipe.lock`, however we would need to ensure the plugin automatically migrated any existing data in composer.json for it to work.

**Comment by tractorcow** [comment]
Actually fixed with https://github.com/silverstripe/recipe-plugin/pull/7, but @chillu has already vetod the idea of a `recipe.lock` file. Unless he wants to modify the A/C we should close this story.
